### ISSUE TITLE
Structure breaks upon undo in Smart Mode

### ISSUE THREAD

**Issue by andreyorst** [justinbarclay/parinfer-rust-mode#10]
From https://github.com/justinbarclay/parinfer-rust-mode/issues/7#issuecomment-642469944

For this input, using smart mode:

```emacs-lisp
(defvar a
  '(:a 1
    :b 2))
```

add space before quote. Smart mode ensures, that below structure is preserved:

```emacs-lisp
(defvar a
   '(:a 1
     :b 2))
```

If we undo once:

```emacs-lisp
(defvar a
   '(:a 1
    :b 2))
```

Structure is broken. Another undo will fix it, but If you've added more than one space structure will remain broken with each undo until final undo. Recording: https://asciinema.org/a/H5NdmLxw0j2Lm9N47z642veLM

I'm using undo-tree package, but it can be reproduced without any undo plugin as well.

Maybe there are some facilities to group modifications into single undo item?

**Comment by andreyorst** [comment]
Maybe take a look at https://github.com/Malabarba/aggressive-indent-mode?

**Comment by andreyorst** [comment]
I've asked at reddit about undo grouping, and got [such reply](https://www.reddit.com/r/emacs/comments/hll3i4/are_there_any_facilities_for_undo_grouping/fx0kzgv/):

```emacs-lisp
(let ((marker
       (prepare-change-group)))
  (unwind-protect (apply fn args)
    (undo-amalgamate-change-group marker)))
```

This looks promising, I'm going to try to integrate this into functions that make changes, so undoing changes would work as single undo.

**Comment by andreyorst** [comment]
Well, good news: it works.

```diff
diff --git a/parinfer-rust-mode.el b/parinfer-rust-mode.el
index b8463ec..f148980 100644
--- a/parinfer-rust-mode.el
+++ b/parinfer-rust-mode.el
@@ -305,27 +305,23 @@ CHANGES."
             (message "%s" (parinfer-rust-print-error error-p)) ;; TODO handle errors
           ;; This stops Emacs from flickering when scrolling
           (if (not (string-equal parinfer-rust--previous-buffer-text replacement-string))
-              (progn
-                (save-mark-and-excursion
-                  (let ((current (current-buffer))
-                        (new-buf (get-buffer-create "*parinfer*")))
-                    (switch-to-buffer new-buf)
-                    (insert replacement-string)
-                    (switch-to-buffer current)
-                    (replace-buffer-contents new-buf)
-                    (kill-buffer new-buf)))
-                ;; Adding an explicit undo boundary here to ensure that when chaining events
-                ;; together that the changes parinfer makes are explicitly recorded and can be
-                ;; undone by themselves. This is meant to help with issues like issue #10, where
-                ;; undo breaks structure and structure is never fully returned to the code during
-                ;; undo. It is left up to the user to realign code.
-                (undo-boundary))))
-        (when-let ((new-x (parinfer-rust-get-in-answer answer "cursor_x"))
-                   (new-line (parinfer-rust-get-in-answer answer "cursor_line")))
-          (parinfer-rust--reposition-cursor new-x new-line))
-        (setq parinfer-rust--previous-options options)
-        (with-no-warnings ;; TODO: Should not need with-no-warnings function
-          (setq-local inhibit-modification-hooks nil))))))
+              (save-mark-and-excursion
+                (let ((change-group (prepare-change-group))
+                      (current (current-buffer))
+                      (new-buf (get-buffer-create "*parinfer*")))
+                  (switch-to-buffer new-buf)
+                  (insert replacement-string)
+                  (switch-to-buffer current)
+                  (replace-buffer-contents new-buf)
+                  (kill-buffer new-buf)
+                  (undo-amalgamate-change-group change-group)
+                  (undo-boundary))))
+          (when-let ((new-x (parinfer-rust-get-in-answer answer "cursor_x"))
+                     (new-line (parinfer-rust-get-in-answer answer "cursor_line")))
+            (parinfer-rust--reposition-cursor new-x new-line))
+          (setq parinfer-rust--previous-options options)
+          (with-no-warnings ;; TODO: Should not need with-no-warnings function
+            (setq-local inhibit-modification-hooks nil)))))))
 
 ;; Interactive or functions meant to be called by user
 (defun parinfer-rust-toggle-debug ()
```

Bad news: it introduces another bug when I hit `C-g` after undo it breaks code horrybly

**Comment by justinbarclay** [comment]
> Maybe take a look how https://github.com/Malabarba/aggressive-indent-mode handles this?

I did scan through there and looked for undo specific code, but unfortunately I didn't find any thing related to undo specifically.

> ... 
> This looks promising, I'm going to try to integrate this into functions that make changes, so undoing changes would work as single undo.

Neat! `undo-amalgamate-changes`  was something on my list to explore with this feature. I chose `undo-boundary` as a stop gap to make sure the user had a way, within the undo paradigm, to maintain the structure of their code (and it was super easy). I don't think it leads to the best experience but it made it more predictable. 🤷‍♀️

> Bad news: it introduces another bug when I hit C-g after undo it breaks code horribly
Does it work better if you remove the `undo-boundary` code, it seems like almagamate and boundary are at odds here?

**Comment by andreyorst** [comment]
> Does it work better if you remove the undo-boundary code, it seems like almagamate and boundary are at odds here?

No. Without `undo-boundary` it works as self insert command, thus if you indent block by 7 spaces this will coun as single undo item. I'm not sure if this is good or bad, but the bugs seem to persist. Although I didn't done much testing.

**Comment by justinbarclay** [comment]
Given the original test case, undoing seems to work well for me. Even with adding a C-g in the middle of a small operation. I'm going to play around with [amalagamate-undo](https://github.com/justinbarclay/parinfer-rust-mode/tree/amalgamate-undo) branch tomorrow and see how well it works for my normal workflow.
https://i.imgur.com/7TkUadA.mp4

**Comment by andreyorst** [comment]
I've used this branch for a bit now, and it seem to work fine, however I was able to break code by using redo. Not sure how to reproduce it... I guess it can be merged, and if I figure our recipe I'll just open new issue. Other than that, it seem to work great.

**Comment by justinbarclay** [comment]
Were you using the redo operation in Emacs 26.1? In 27.x (redo-only, undo-only)? Or in Undo-tree?

My thoughts on this right now is that I'm fine siding on either undo-amalgamate or undo-boundary. 

I think undo-amalgamate is nice because it creates a nice user experience when it works and limits the amount of undo one has to perform to get to their intended state.

I think undo-boundary is nice because there are less edge cases around using it, it supports older emacs versions, and it's (almost) guaranteed to leave your buffer in a reasonable state based on your actions. The downside is it becomes two undo operations per command, so there is more frustration when using it.

**Comment by andreyorst** [comment]
> Were you using the redo operation in Emacs 26.1? In 27.x (redo-only, undo-only)? Or in Undo-tree?

I'm using Emacs 28.0.50 compiled from `feature/native-comp` branch, as well as one compiled from `master` branch on other machine, and `undo-tree` package.

> My thoughts on this right now is that I'm fine siding on either undo-amalgamate or undo-boundary.
>
>I think undo-amalgamate is nice because it creates a nice user experience when it works and limits the amount of undo one has to perform to get to their intended state.
>
>I think undo-boundary is nice because there are less edge cases around using it, it supports older emacs versions, and it's (almost) guaranteed to leave your buffer in a reasonable state based on your actions. The downside is it becomes two undo operations per command, so there is more frustration when using it.

You can use both, It'll work. But there's the bug I was talking about.

However I feel that using only `undo-amalgamate` is better, because it works **as** it works in buffers that do not use Parinfer, so Parinfer just keeps indentation in sync, while being as transparent as possible. And it seem to work better than with `undo-boundary` since there's no structure breakage.

So I'm all for amalgamating

> it supports older emacs versions

Ahem, [requirements](https://github.com/justinbarclay/parinfer-rust-mode#requirements), ahem, Emacs 26+. And amalgamate function is:

>`(undo-amalgamate-change-group HANDLE)`
>
>  Probably introduced at or before Emacs version 26.1.

So I guess all Emacs 26+ users are covered :)

**Comment by justinbarclay** [comment]
I share your thoughts on amalgamate being a good first choice. It’s definitely a better UX and fits better in the philosophy of Emacs more. It will probably be merged sometime this weekend as I’ve got some more lisp development to do today, allowing me to test it some more.

As an aside...

To me Emacs 26 is old as well, but 26.3 is the latest stable release. And IIRC Emacs 25.3 is the default version that can be installed with apt.

27 and 28 are still pretty out there for a lot of Emacs users. I mean, 27 is still getting ready to be a release candidate and 28 is bleeding edge.

Emacs 25 was the first version to get dynamic module, so that’s the lowest bar I can shoot for. I know I only support Emacs 26+ due to the functions I use, so undo-amalgamate doesn’t change anything there. But my goal has always been to eventually support the 2 newest stable versions of Emacs. (And if I wait long enough I’ll get that for free when Emacs 27 is released as stable :p )

**Comment by andreyorst** [comment]
I guess you can do comptime branching on version to use either amalgamation or boundaries?

**Comment by justinbarclay** [comment]
Yeah like I said I’m not there yet, I’ve got bigger fish to fry for this project first :)

However, before I did any load/compile time swapping I would try to backport the function. (As long as I don’t have to pull along other functions in the process)

**Comment by andreyorst** [comment]
> Emacs 25 was the first version to get dynamic module, so that’s the lowest bar I can shoot for. I know I only support Emacs 26+ due to the functions I use, so undo-amalgamate doesn’t change anything there. But my goal has always been to eventually support the 2 newest stable versions of Emacs. (And if I wait long enough I’ll get that for free when Emacs 27 is released as stable :p )

Also, I was developing a bunch of plugins for Vim and Kakoune, and for Vim I've tried to be compatible with Vim version from Ubuntu repos, and gave up pretty quickly because of how outdated it was at the time. For Kakoune I've actually supported latest stable and master branch, which were pretty different due to pace at which Kakoune is developed. And gave up again because it made the code very clunky and required to duplicate code but throw away new features that make code shorter. Maybe Emacs is a different story though - this is the first package I've ever contributed elisp code to so I don't know :)

What I mean is that lowest bar is not always an ideal target and often can lead to bad code hanging here and there iinstead of clean code base.

> I would try to backport the function. (As long as I don’t have to pull along other functions in the process)

Given that this is an ordinary lisp function it seems it can be backported though, and function itself doesn't seem to use any other functions, so maybe backporting it will be sufficient. Perhaps no one needed it before that's why it appeared recently, e.g. there should be no technical difficulties for Emacs 25 to support it, since it automatically amalgamates undos.
